### ISSUE TITLE
Date filters throw error (BigQuery, v0.33.4)

### ISSUE THREAD

**Issue by shafy** [metabase/metabase#11222]
**Describe the bug**
When using the field type `DATETIME` in BigQuery, I get the following errors when filtering in the GUI. It doesn't happen with when using the field type `TIMESTAMP` with the exact data. I think it happens only with the "on" and "between" date filter choices.

```
No matching signature for operator >= for argument types: DATETIME, TIMESTAMP. Supported signatures: ANY >= ANY at [2:825]
```

Here's the SQL:

```
SELECT `simplified_prod.rentals`.`rental_id` AS `rental_id`, `simplified_prod.rentals`.`battery_id` AS `battery_id`, `simplified_prod.rentals`.`city` AS `city`, `simplified_prod.rentals`.`created_at` AS `created_at`, `simplified_prod.rentals`.`external_identifier` AS `external_identifier`, `simplified_prod.rentals`.`latitude` AS `latitude`, `simplified_prod.rentals`.`location_id` AS `location_id`, `simplified_prod.rentals`.`location_name` AS `location_name`, `simplified_prod.rentals`.`longitude` AS `longitude`, `simplified_prod.rentals`.`partner_name` AS `partner_name`, `simplified_prod.rentals`.`pos_number` AS `pos_number`, `simplified_prod.rentals`.`region` AS `region`, `simplified_prod.rentals`.`rental_type` AS `rental_type`, `simplified_prod.rentals`.`status` AS `status`
FROM `simplified_prod.rentals`
WHERE (`simplified_prod.rentals`.`created_at` >= timestamp('2019-10-28T00:00:00.000Z')
   AND `simplified_prod.rentals`.`created_at` < timestamp('2019-10-29T00:00:00.000Z'))
LIMIT 1048576
```

**Logs**
```
{:status :failed,
 :class clojure.lang.ExceptionInfo,
 :error "No matching signature for operator >= for argument types: DATETIME, TIMESTAMP. Supported signatures: ANY >= ANY at [2:825]",
 :stacktrace
 ("--> driver.google$execute_no_auto_retry.invokeStatic(google.clj:39)"
  "driver.google$execute_no_auto_retry.invoke(google.clj:32)"
  "driver.google$execute$fn__367.invoke(google.clj:50)"
  "util$do_with_auto_retries.invokeStatic(util.clj:429)"
  "util$do_with_auto_retries.invoke(util.clj:421)"
  "util$do_with_auto_retries.invokeStatic(util.clj:433)"
  "util$do_with_auto_retries.invoke(util.clj:421)"
  "util$do_with_auto_retries.invokeStatic(util.clj:433)"
  "util$do_with_auto_retries.invoke(util.clj:421)"
  "driver.google$execute.invokeStatic(google.clj:49)"
  "driver.google$execute.invoke(google.clj:43)"
  "driver.bigquery$execute_bigquery.invokeStatic(bigquery.clj:204)"
  "driver.bigquery$execute_bigquery.invoke(bigquery.clj:193)"
  "driver.bigquery$execute_bigquery.invokeStatic(bigquery.clj:195)"
  "driver.bigquery$execute_bigquery.invoke(bigquery.clj:193)"
  "driver.bigquery$process_native_STAR_$fn__925.invoke(bigquery.clj:211)"
  "util$do_with_auto_retries.invokeStatic(util.clj:429)"
  "util$do_with_auto_retries.invoke(util.clj:421)"
  "util$do_with_auto_retries.invokeStatic(util.clj:433)"
  "util$do_with_auto_retries.invoke(util.clj:421)"
  "driver.bigquery$process_native_STAR_.invokeStatic(bigquery.clj:210)"
  "driver.bigquery$process_native_STAR_.invoke(bigquery.clj:206)"
  "driver.bigquery$eval931$fn__933.invoke(bigquery.clj:226)"
  "query_processor$fn__44449$execute_query__44454$fn__44455.invoke(query_processor.clj:71)"
  "query_processor$fn__44449$execute_query__44454.invoke(query_processor.clj:65)"
  "query_processor.middleware.mbql_to_native$mbql__GT_native$fn__20887.invoke(mbql_to_native.clj:38)"
  "query_processor.middleware.annotate$result_rows_maps__GT_vectors$fn__36504.invoke(annotate.clj:541)"
  "query_processor.middleware.annotate$add_column_info$fn__36410.invoke(annotate.clj:485)"
  "query_processor.middleware.cumulative_aggregations$handle_cumulative_aggregations$fn__37445.invoke(cumulative_aggregations.clj:57)"
  "query_processor.middleware.resolve_joins$resolve_joins$fn__41557.invoke(resolve_joins.clj:184)"
  "query_processor.middleware.limit$limit$fn__38522.invoke(limit.clj:19)"
  "query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__44316.invoke(results_metadata.clj:87)"
  "query_processor.middleware.format_rows$format_rows$fn__38510.invoke(format_rows.clj:26)"
  "query_processor.middleware.add_dimension_projections$add_remapping$fn__34938.invoke(add_dimension_projections.clj:234)"
  "query_processor.middleware.add_source_metadata$add_source_metadata_for_source_queries$fn__35638.invoke(add_source_metadata.clj:107)"
  "query_processor.middleware.resolve_source_table$resolve_source_tables$fn__41607.invoke(resolve_source_table.clj:46)"
  "query_processor.middleware.add_row_count_and_status$add_row_count_and_status$fn__35498.invoke(add_row_count_and_status.clj:16)"
  "query_processor.middleware.driver_specific$process_query_in_context$fn__37521.invoke(driver_specific.clj:12)"
  "query_processor.middleware.resolve_driver$resolve_driver$fn__41221.invoke(resolve_driver.clj:22)"
  "query_processor.middleware.bind_effective_timezone$bind_effective_timezone$fn__36835$fn__36836.invoke(bind_effective_timezone.clj:9)"
  "util.date$call_with_effective_timezone.invokeStatic(date.clj:88)"
  "util.date$call_with_effective_timezone.invoke(date.clj:77)"
  "query_processor.middleware.bind_effective_timezone$bind_effective_timezone$fn__36835.invoke(bind_effective_timezone.clj:8)"
  "query_processor.middleware.store$initialize_store$fn__44341$fn__44342.invoke(store.clj:11)"
  "query_processor.store$do_with_store.invokeStatic(store.clj:46)"
  "query_processor.store$do_with_store.invoke(store.clj:40)"
  "query_processor.middleware.store$initialize_store$fn__44341.invoke(store.clj:10)"
  "query_processor.middleware.async$async__GT_sync$fn__20798.invoke(async.clj:23)"
  "query_processor.middleware.async_wait$runnable$fn__36561.invoke(async_wait.clj:89)"),
 :query
 {:query {:source-table 5, :filter ["=" ["field-id" 43] "2019-10-28"]},
  :type "query",
  :parameters [],
  :async? true,
  :middleware {:add-default-userland-constraints? true, :userland-query? true},
  :info
  {:executed-by 2,
   :context :ad-hoc,
   :card-id nil,
   :nested? false,
   :query-hash [77, -116, 115, 72, -8, -67, -24, -66, 103, -115, 94, -9, 98, 100, 27, -8, 77, 7, 60, -61, -5, -62, 50, -3, -90, -4, 38, -100, 111, 8, -29, 71]},
  :constraints {:max-results 10000, :max-results-bare-rows 2000}},
 :preprocessed
 {:constraints {:max-results 10000, :max-results-bare-rows 2000},
  :type :query,
  :middleware {:add-default-userland-constraints? true, :userland-query? true},
  :info
  {:executed-by 2,
   :context :ad-hoc,
   :nested? false,
   :query-hash [77, -116, 115, 72, -8, -67, -24, -66, 103, -115, 94, -9, 98, 100, 27, -8, 77, 7, 60, -61, -5, -62, 50, -3, -90, -4, 38, -100, 111, 8, -29, 71]},
  :preprocessing-level 1,
  :database 2,
  :driver :bigquery,
  :query
  {:source-table 5,
   :filter
   [:and
    [:>= [:datetime-field [:field-id 43] :default] [:absolute-datetime #inst "2019-10-28T00:00:00.000000000-00:00" :default]]
    [:< [:datetime-field [:field-id 43] :default] [:absolute-datetime #inst "2019-10-29T00:00:00.000000000-00:00" :default]]],
   :fields
   [[:field-id 50]
    [:field-id 41]
    [:field-id 44]
    [:datetime-field [:field-id 43] :default]
    [:field-id 48]
    [:field-id 42]
    [:field-id 39]
    [:field-id 37]
    [:field-id 47]
    [:field-id 38]
    [:field-id 40]
    [:field-id 45]
    [:field-id 46]
    [:field-id 49]],
   :limit 2000}},
 :native
 {:query
  "SELECT `simplified_prod.rentals`.`rental_id` AS `rental_id`, `simplified_prod.rentals`.`battery_id` AS `battery_id`, `simplified_prod.rentals`.`city` AS `city`, `simplified_prod.rentals`.`created_at` AS `created_at`, `simplified_prod.rentals`.`external_identifier` AS `external_identifier`, `simplified_prod.rentals`.`latitude` AS `latitude`, `simplified_prod.rentals`.`location_id` AS `location_id`, `simplified_prod.rentals`.`location_name` AS `location_name`, `simplified_prod.rentals`.`longitude` AS `longitude`, `simplified_prod.rentals`.`partner_name` AS `partner_name`, `simplified_prod.rentals`.`pos_number` AS `pos_number`, `simplified_prod.rentals`.`region` AS `region`, `simplified_prod.rentals`.`rental_type` AS `rental_type`, `simplified_prod.rentals`.`status` AS `status` FROM `simplified_prod.rentals` WHERE (`simplified_prod.rentals`.`created_at` >= timestamp('2019-10-28T00:00:00.000Z') AND `simplified_prod.rentals`.`created_at` < timestamp('2019-10-29T00:00:00.000Z')) LIMIT 2000",
  :table-name "rentals",
  :mbql? true},
 :ex-data
 {"code" 400,
  "errors"
  [{"domain" "global", "location" "q", "locationType" "parameter", "message" "No matching signature for operator >= for argument types: DATETIME, TIMESTAMP. Supported signatures: ANY >= ANY at [2:825]", "reason" "invalidQuery"}],
  "message" "No matching signature for operator >= for argument types: DATETIME, TIMESTAMP. Supported signatures: ANY >= ANY at [2:825]",
  "status" "INVALID_ARGUMENT"}}
```

**Information about your Metabase Installation:**
```
{
  "browser-info": {
    "language": "en-US",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.14; rv:69.0) Gecko/20100101 Firefox/69.0",
    "vendor": ""
  },
  "system-info": {
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "1.8.0_232-heroku-b09",
    "java.vendor": "Oracle Corporation",
    "java.vendor.url": "http://java.oracle.com/",
    "java.version": "1.8.0_232-heroku",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "25.232-b09",
    "os.name": "Linux",
    "os.version": "4.4.0-1048-aws",
    "user.language": "en",
    "user.timezone": "Etc/UTC"
  },
  "metabase-info": {
    "databases": [
      "bigquery"
    ],
    "hosting-env": "heroku",
    "application-database": "postgres",
    "run-mode": "prod",
    "version": {
      "tag": "v0.33.4",
      "date": "2019-10-07",
      "branch": "release-0.33.x",
      "hash": "9559406"
    },
    "settings": {
      "report-timezone": null
    }
  }
}
```

**Severity**
Semi-annoying, as now I have to convert my fields two `TIMESTAMP` in BigQuery

**Comment by j-manu** [comment]
Encountering the same problem. Seems to be a regression

- https://github.com/metabase/metabase/issues/11222
- https://github.com/metabase/metabase/issues/8932

**Comment by camsaul** [comment]
Most other SQL databases are smart enough to know how to compare a `datetime` and `timestamp` automatically but it looks like BigQuery isn't unfortunately. So in this case we need to generate datetime literals instead of `timestamp(...)` ones. That should be fairly easy to fix. I'll take a look

**Comment by mazameli** [comment]
@camsaul It would be cool to get a fix for this into 0.34, especially if it's `fairly easy to fix`.
